File: app/src/main/java/com/chooloo/www/callmanager/ui/search/SearchFragment.java
package com.chooloo.www.callmanager.ui.search;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;

import com.chooloo.www.callmanager.R;
import com.chooloo.www.callmanager.databinding.FragmentSearchBarBinding;
import com.chooloo.www.callmanager.ui.base.BaseFragment;

public class SearchFragment extends BaseFragment implements SearchMvpView {

    OnFocusChangedListener mOnFocusChangedListener;
    OnTextChangedListener mOnTextChangedListener;

    private SearchMvpPresenter<SearchMvpView> mPresenter;
    private FragmentSearchBarBinding binding;

    public static SearchFragment newInstance() {
        Bundle args = new Bundle();
        SearchFragment fragment = new SearchFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentSearchBarBinding.inflate(inflater);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mPresenter.attach(this);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mPresenter.detach();
    }

    @Override
    public void onSetup() {
        mPresenter = new SearchPresenter<>();
        mPresenter.attach(this);

        mOnFocusChangedListener = isFocused -> {
        };
        mOnTextChangedListener = text -> {
        };

        binding.searchInputEditText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                mPresenter.onTextChanged(String.valueOf(s));
                mOnTextChangedListener.onTextChanged(String.valueOf(s));
            }

            @Override
            public void afterTextChanged(Editable s) {
            }
        });
        binding.searchInputEditText.setOnFocusChangeListener((v, hasFocus) -> {
            mPresenter.onFocusChanged(hasFocus);
            mOnFocusChangedListener.onFocusChanged(hasFocus);
        });
    }

    @Override
    public String getText() {
        return binding.searchInputEditText.getText().toString();
    }

    @Override
    public void setFocus() {
        binding.searchInputEditText.requestFocus();
    }

    @Override
    public void showIcon(boolean isShow) {
//        binding.searchInputEditText.icon
//        binding.searchInputLayout.setStartIconVisible(isShow);
        binding.searchInputEditText.setCompoundDrawablesWithIntrinsicBounds(isShow ? ContextCompat.getDrawable(_activity, R.drawable.ic_search_black_24dp) : null, null, null, null);
    }

    public void setOnFocusChangedListener(OnFocusChangedListener onFocusChangedListener) {
        this.mOnFocusChangedListener = onFocusChangedListener;
    }

    public void setOnTextChangedListener(OnTextChangedListener onTextChangedListener) {
        this.mOnTextChangedListener = onTextChangedListener;
    }

    public interface OnFocusChangedListener {
        void onFocusChanged(boolean isFocused);
    }

    public interface OnTextChangedListener {
        void onTextChanged(String text);
    }
}

File: app/src/main/java/com/chooloo/www/callmanager/ui/search/SearchFragment.kt
package com.chooloo.www.callmanager.ui.search

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.core.widget.addTextChangedListener
import com.chooloo.www.callmanager.R
import com.chooloo.www.callmanager.databinding.FragmentSearchBarBinding
import com.chooloo.www.callmanager.ui.base.BaseFragment

class SearchFragment : BaseFragment(), SearchMvpView {

    private var _onFocusChangedListener: ((isFocused: Boolean) -> Unit?)? = null
    private var _onTextChangedListener: ((text: String) -> Unit?)? = null
    private lateinit var _presenter: SearchMvpPresenter<SearchMvpView>
    private lateinit var _binding: FragmentSearchBarBinding

    companion object {
        fun newInstance(): SearchFragment = SearchFragment()
    }

    override val text: String
        get() = _binding.searchInputEditText.text.toString()


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        _binding = FragmentSearchBarBinding.inflate(inflater)
        return _binding.root
    }

    override fun onSetup() {
        _presenter = SearchPresenter()
        _presenter.attach(this)

        _binding.apply {
            searchInputEditText.addTextChangedListener(
                    afterTextChanged = {},
                    onTextChanged = { _, _, _, _ -> },
                    beforeTextChanged = { s, _, _, _ ->
                        _presenter.onTextChanged(s.toString())
                        _onTextChangedListener?.invoke(s.toString())
                    }
            )
            searchInputEditText.setOnFocusChangeListener { _, hasFocus ->
                run {
                    _presenter.onFocusChanged(hasFocus)
                    _onFocusChangedListener?.invoke(hasFocus)
                }
            }
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _presenter.detach()
    }

    override fun setFocus() {
        _binding.searchInputEditText.requestFocus()
    }

    override fun showIcon(isShow: Boolean) {
        _binding.searchInputEditText.setCompoundDrawablesWithIntrinsicBounds(if (isShow) ContextCompat.getDrawable(_activity, R.drawable.ic_search_black_24dp) else null, null, null, null)
    }

    fun setOnFocusChangedListener(onFocusChangedListener: ((isFocused: Boolean) -> Unit?)?) {
        _onFocusChangedListener = onFocusChangedListener
    }

    fun setOnTextChangedListener(onTextChangedListener: ((text: String) -> Unit?)?) {
        _onTextChangedListener = onTextChangedListener
    }
}